app/models/listing.rb

```diff
@@ -34,6 +34,26 @@ class Listing < ApplicationRecord
 
   ALLOWED_STATES = ["AL", "AK", "AZ", "AR", "CA", "CO", "CT", "DE", "FL", "GA", "HI", "ID", "IL", "IN", "IA", "KS", "KY", "LA", "ME", "MD", "MA", "MI", "MN", "MS", "MO", "MT", "NE", "NV", "NH", "NJ", "NM", "NY", "NC", "ND", "OH", "OK", "OR", "PA", "RI", "SC", "SD", "TN", "TX", "UT", "VT", "VA", "WA", "WV", "WI", "WY"].freeze
 
+  def self.full_search(search_options)
+    name_query = search_options[:name]
+    location_query = search_options[:location]
+    category_query = search_options[:category]
+    coordinates_query = search_options[:coordinates]
+
+    if name_query && location_query
+      near(location_query, 5).search_by_name(name_query)
+    elsif name_query
+      search_by_name(name_query)
+    elsif location_query
+      near(location_query, 5)
+    elsif category_query
+      categories = Category.search_by_name(category_query)
+      categories.map(&:listings).flatten.uniq
+    elsif coordinates_query
+      near(coordinates_query, 5)
+    end
+  end
```

Comment:
@dladowitz here's one idea:

```ruby
SEARCHES = {
  name: ->(scope, term) { scope.search_by_name(term) },
  location: ->(scope, term) { scope.near(term, 5) },
  category: ->(scope, term) { scope.merge(Category.search_by_name(term)) },
  coordinates: ->(scope, term) { scope.near(term, 5) },
}.freeze

def self.full_search(search_options)
  scope = all

  search_options.each do |search_type, term|
    scope = SEARCHES.fetch(search_type).call(scope, term)
  end

  scope
end
```

Not sure if the `.merge` bit will work as-is, might need some finagling. The upshot of this approach is that it allows us to apply arbitrary searches together.

Reply:
Yeah, I like this. I can't get it to work though. 
The main issue is that Category.search_by_name(term) returns a relation of Categories and it needs to return a relation of Listings. I could only get it to return an array of Listings. 

Feel free to try and make it work.